### PR TITLE
Update parameters to increase the accuracy

### PR DESCRIPTION
**Pull request by sreenikSS**

### CODE REVIEW

File: Kaggle/DigitRecognizer/src/DigitRecognizer.cpp [hunk 1]
```diff
@@ -75,10 +73,14 @@ int main()
...
   data::Split(dataset, train, valid, RATIO);
 
-  // Getting training and validating dataset with features only.
-  const mat trainX = train.submat(1, 0, train.n_rows - 1, train.n_cols - 1);
-  const mat validX = valid.submat(1, 0, valid.n_rows - 1, valid.n_cols - 1);
-
+  // Getting training and validating dataset with features only and then
+  // normalising
+  const mat trainX = train.submat(1, 0, train.n_rows - 1,
+      train.n_cols - 1)/255.0;
```
Comment (zoq) [inline Kaggle/DigitRecognizer/src/DigitRecognizer.cpp]: Missing space between the divide operation.

File: Kaggle/DigitRecognizer/src/DigitRecognizer.cpp [hunk 2]
```diff
@@ -75,10 +73,14 @@ int main()
...
+  const mat validX = valid.submat(1, 0, valid.n_rows - 1,
+      valid.n_cols - 1)/255.0;
```
Comment (zoq) [inline Kaggle/DigitRecognizer/src/DigitRecognizer.cpp]: See comment above.

File: Kaggle/DigitRecognizer/src/DigitRecognizer.cpp [hunk 3]
```diff
@@ -34,28 +35,25 @@ using namespace ens;
 int main()
 {
   // Dataset is randomly split into validation
-  // and training parts with following ratio.
+  // and training parts in the following ratio.
   constexpr double RATIO = 0.1;
   // The number of neurons in the first layer.
-  constexpr int H1 = 100;
+  constexpr int H1 = 200;
```
Comment (zoq) [inline Kaggle/DigitRecognizer/src/DigitRecognizer.cpp]: Not sure if it is necessary to increase the hidden layer size?

Comment (sreenikSS) [reply Kaggle/DigitRecognizer/src/DigitRecognizer.cpp]: Well I did experiment with different sizes. Changing the first hidden layer size to 200 brings upon a slight improvement while increasing it further does not help. Shall I keep it as 200 or revert back to 100?

Comment (rcurtin) [reply Kaggle/DigitRecognizer/src/DigitRecognizer.cpp]: Personally I think 200 is fine.

**Comment by sreenikSS** [comment]
@rcurtin I did a bit of an analysis here which I think is interesting to observe and can be of help to anyone new.

## Accuracy after all changes

1 - accuracy: train = 95.4524%, valid = 94.9286%
2 - accuracy: train = 96.9974%, valid = 96.0952%
3 - accuracy: train = 97.6481%, valid = 96.5%
4 - accuracy: train = 98.0079%, valid = 96.5476%
5 - accuracy: train = 98.4259%, valid = 97.0238%
6 - accuracy: train = 97.5767%, valid = 96.0714%
7 - accuracy: train = 98.8307%, valid = 97.3571%
8 - accuracy: train = 97.9074%, valid = 96.2857%
9 - accuracy: train = 98.2037%, valid = 96.5952%
10 - accuracy: train = 98.0132%, valid = 96.0238%
11 - accuracy: train = 98.3836%, valid = 96.5%
12 - accuracy: train = 98.6772%, valid = 96.9286%
13 - accuracy: train = 98.9577%, valid = 97.0714%
14 - accuracy: train = 98.5608%, valid = 96.4048%
15 - accuracy: train = 98.7963%, valid = 96.5952%
16 - accuracy: train = 99.1958%, valid = 97.0714%
17 - accuracy: train = 99.2037%, valid = 97%
18 - accuracy: train = 99.3042%, valid = 97.2143%
19 - accuracy: train = 99.2116%, valid = 97%
20 - accuracy: train = 99.1931%, valid = 97.0714%

Now let's start from the initial model and tune the parameters few at a time.
#### Note: The step size is adjusted accordingly in each step to limit the cycles to 20.

## Starting accuracy

1 - accuracy: train = 29.7196%, valid = 28.5238%
2 - accuracy: train = 46.2354%, valid = 46.6667%
3 - accuracy: train = 56.1058%, valid = 56.5%
4 - accuracy: train = 62.6402%, valid = 62.4286%
5 - accuracy: train = 66.6878%, valid = 66.7619%
6 - accuracy: train = 69.582%, valid = 69.6905%
7 - accuracy: train = 71.836%, valid = 71.2381%
8 - accuracy: train = 73.8492%, valid = 73.2143%
9 - accuracy: train = 75.4709%, valid = 74.3095%
10 - accuracy: train = 76.3757%, valid = 75.8333%
11 - accuracy: train = 77.7751%, valid = 77.119%
12 - accuracy: train = 78.3386%, valid = 77.7857%
13 - accuracy: train = 79.6667%, valid = 78.8571%
14 - accuracy: train = 80.0556%, valid = 79.4048%
15 - accuracy: train = 80.746%, valid = 80.5238%
16 - accuracy: train = 81.0635%, valid = 80.881%
17 - accuracy: train = 81.4815%, valid = 81.7619%
18 - accuracy: train = 82.1825%, valid = 82.0714%
19 - accuracy: train = 82.6455%, valid = 82.3095%
20 - accuracy: train = 82.9868%, valid = 82.2143%

## Changing only the network architecture (replacing sigmoid with relu, adding dropout and other minor changes)

1 - accuracy: train = 84.4021%, valid = 84.6667%
2 - accuracy: train = 88.5926%, valid = 88.0238%
3 - accuracy: train = 90.5556%, valid = 89.9286%
4 - accuracy: train = 92.0344%, valid = 90.4048%
5 - accuracy: train = 93.1058%, valid = 91.5476%
6 - accuracy: train = 93.8466%, valid = 91.881%
7 - accuracy: train = 94.6852%, valid = 92.2619%
8 - accuracy: train = 95.1905%, valid = 92.7619%
9 - accuracy: train = 95.8677%, valid = 92.8571%
10 - accuracy: train = 96.1852%, valid = 92.9286%
11 - accuracy: train = 96.4101%, valid = 93.119%
12 - accuracy: train = 96.9153%, valid = 93.2143%
13 - accuracy: train = 97.0688%, valid = 93.3333%
14 - accuracy: train = 97.4709%, valid = 93.5238%
15 - accuracy: train = 97.7196%, valid = 93.9048%
16 - accuracy: train = 97.7063%, valid = 93.6667%
17 - accuracy: train = 98.0926%, valid = 93.881%
18 - accuracy: train = 98.2857%, valid = 93.9524%
19 - accuracy: train = 98.3889%, valid = 93.9048%
20 - accuracy: train = 98.5238%, valid = 94.2143%

The model is clearly overfitting and increasing the dropout does not help here.

The only changes left to be done are the init type and normalization.

## Normalizing the data (still has random initialization)

1 - accuracy: train = 88.3889%, valid = 88.0714%
2 - accuracy: train = 89.9974%, valid = 88.9762%
3 - accuracy: train = 91.3201%, valid = 89.619%
4 - accuracy: train = 92.7381%, valid = 90.5714%
5 - accuracy: train = 94.1508%, valid = 91.1429%
6 - accuracy: train = 95.0238%, valid = 92.0952%
7 - accuracy: train = 95.5106%, valid = 92.381%
8 - accuracy: train = 95.8095%, valid = 92.619%
9 - accuracy: train = 95.8016%, valid = 92.8333%
10 - accuracy: train = 96.8069%, valid = 93.619%
11 - accuracy: train = 96.4841%, valid = 93.4524%
12 - accuracy: train = 96.9444%, valid = 93.5476%
13 - accuracy: train = 96.7725%, valid = 93.7143%
14 - accuracy: train = 97.2407%, valid = 94.5476%
15 - accuracy: train = 97.5608%, valid = 94.4762%
16 - accuracy: train = 97.4815%, valid = 94.9524%
17 - accuracy: train = 97.6429%, valid = 95.119%
18 - accuracy: train = 97.7249%, valid = 95.0952%
19 - accuracy: train = 97.8757%, valid = 95.2619%
20 - accuracy: train = 98.2884%, valid = 95.4524%

## Changing the initialization type to glorot (but not normalizing the data)

1 - accuracy: train = 95.2513%, valid = 94.5238%
2 - accuracy: train = 97.0291%, valid = 96.3333%
3 - accuracy: train = 96.873%, valid = 96.0476%
4 - accuracy: train = 97.0847%, valid = 96.119%
5 - accuracy: train = 96.8624%, valid = 95.5476%
6 - accuracy: train = 97.0794%, valid = 95.881%
7 - accuracy: train = 97.6693%, valid = 96.2381%
8 - accuracy: train = 96.4286%, valid = 95.0476%
9 - accuracy: train = 97.4392%, valid = 95.6667%
10 - accuracy: train = 97.7434%, valid = 95.8571%
11 - accuracy: train = 97.2116%, valid = 95.881%
12 - accuracy: train = 98.4683%, valid = 96.3571%
13 - accuracy: train = 98.6323%, valid = 96.9524%
14 - accuracy: train = 98.6958%, valid = 96.8333%
15 - accuracy: train = 98.2619%, valid = 96.3571%
16 - accuracy: train = 98.619%, valid = 96.7381%
17 - accuracy: train = 98.3254%, valid = 96.6667%
18 - accuracy: train = 98.6402%, valid = 96.7143%
19 - accuracy: train = 99.1349%, valid = 97.0714%
20 - accuracy: train = 98.254%, valid = 96.5714%

This might be a little strange but it explains the fact that a limited range like 0-255 is easily learn-able without initial data normalization. Batch normalization should help against overfitting.

Finally I also observed the raw weights and biases with and without normalization and they seemed to be more or less of the same order with the parameters of the former case having slightly (almost negligibly) larger values.

**Comment by rcurtin** [comment]
I don't have mlpack-bot set up on this repository but it would have approved by now, so I'll go ahead and merge.  Thanks again. :+1:
